tests/serializers/test_data.py

```diff
@@ -420,70 +405,98 @@ def inherited_compare(testcase, pk, klass, data):
+def assert_serializer(self, format, data):
+    # Create all the objects defined in the test data.
     objects = []
-    instance_count = {}
-    for func, pk, klass, datum in test_data:
+    for test_helper, pk, model, data_value in data:
         with connection.constraint_checks_disabled():
-            objects.extend(func[0](pk, klass, datum))
+            objects.extend(test_helper.create_object(pk, model, data_value))
 
-    # Get a count of the number of objects created for each class
-    for klass in instance_count:
-        instance_count[klass] = klass.objects.count()
+    # Get a count of the number of objects created for each model class.
+    instance_counts = {}
+    for _, _, model, _ in data:
+        if model not in instance_counts:
+            instance_counts[model] = model.objects.count()
 
-    # Add the generic tagged objects to the object list
+    # Add the generic tagged objects to the object list.
     objects.extend(Tag.objects.all())
 
-    # Serialize the test database
+    # Serialize the test database.
     serialized_data = serializers.serialize(format, objects, indent=2)
 
     for obj in serializers.deserialize(format, serialized_data):
         obj.save()
 
-    # Assert that the deserialized data is the same
-    # as the original source
-    for func, pk, klass, datum in test_data:
-        func[1](self, pk, klass, datum)
+    # Assert that the deserialized data is the same as the original source.
+    for test_helper, pk, model, data_value in data:
+        with self.subTest(model=model, data_value=data_value):
+            test_helper.compare_object(self, pk, model, data_value)
+
+    # Assert no new objects were created.
+    for model, count in instance_counts.items():
+        with self.subTest(model=model, count=count):
+            self.assertEqual(count, model.objects.count())
 
-    # Assert that the number of objects deserialized is the
-    # same as the number that was serialized.
-    for klass, count in instance_count.items():
-        self.assertEqual(count, klass.objects.count())
+
+def serializerTest(self, format):
+    assert_serializer(self, format, test_data)
+
+
+@skipUnlessDBFeature("allows_auto_pk_0")
+def serializerTestPK0(self, format):
+    # FK to an object with PK of 0. This won't work on MySQL without the
+    # NO_AUTO_VALUE_ON_ZERO SQL mode since it won't let you create an object
+    # with an autoincrement primary key of 0.
+    data = [
+        (data_obj, 0, Anchor, "Anchor 0"),
+        (fk_obj, 1, FKData, 0),
+    ]
+    assert_serializer(self, format, data)
+
+
+@skipIfDBFeature("interprets_empty_strings_as_nulls")
+def serializerTestEmptyString(self, format):
+    data = [
+        (data_obj, 1, BinaryData, None),
+        (data_obj, 2, CharData, None),
+        (data_obj, 3, EmailData, None),
+        (data_obj, 4, FilePathData, None),
+        (data_obj, 5, SlugData, None),
+        (data_obj, 6, TextData, None),
+        (data_obj, 7, FileData, None),
```

Comment:
@sarahboyce I believe the tests are failing because this line should be commented out. It's one of the commented-out cases that I'm going to try to address to close the ticket.

https://github.com/django/django/blob/3d819e23240ab6f903eb60d76237628981fd52b5/tests/serializers/test_data.py#L280

Reply:
Thank you